datafusion/datasource/src/source.rs

```diff
@@ -79,8 +82,20 @@ pub trait DataSource: Send + Sync + Debug {
         &self,
         _projection: &ProjectionExec,
     ) -> datafusion_common::Result<Option<Arc<dyn ExecutionPlan>>>;
+    /// Push down filters from parent execution plans to this data source.
+    /// This is expected to return Ok(None) if the filters cannot be pushed down.
+    /// If they can be pushed down it should return a [`FilterPushdownResult`] containing the new
+    /// data source and the support level for each filter (exact or inexact).
+    fn push_down_filters(
+        &self,
+        _filters: &[PhysicalExprRef],
+    ) -> datafusion_common::Result<Option<DataSourceFilterPushdownResult>> {
```

Comment:
Yeah, this type is getting pretty complicated. Maybe we could add the `Option` to `FilterPushdownResult`-- maybe make it a enum with variants for did pushdown, or no pushdown

```rust
pub enum FilterPushdownResult<T> {
    Pushed {
      pub inner: T,
      pub support: Vec<FilterSupport>,
    },
    /// filters could not be pushed down
    NotPushed
}
```

Reply:
Yep makes sense, happy to do that. The important bit is that you can cheaply return the `NotPushed` variant without having to clone yourself, etc.